src/metabase/query_processor/middleware/annotate.clj

```diff
@@ -93,16 +93,17 @@
    rf
    [(base-type-inferer metadata)]
    (fn combine [result base-types]
-     (let [cols' (mapv (fn [col base-type]
-                         (assoc col
-                                :base_type      base-type
-                                :effective_type base-type
-                                :field_ref      [:field (:name col) {:base-type base-type}]))
-                       (:cols metadata)
-                       base-types)]
+     (let [incoming-cols (when (map? result)
+                           (-> result :data :cols))]
        (rf (cond-> result
-             (map? result)
-             (assoc-in [:data :cols] cols')))))))
+             (= (count incoming-cols) (count base-types))
+             (assoc-in [:data :cols] (mapv (fn [col base-type]
+                                             (assoc col
+                                                    :base_type      base-type
+                                                    :effective_type base-type
+                                                    :field_ref      [:field (:name col) {:base-type base-type}]))
```

Comment:
I know this is what the old code was doing but it would probably be safer just to update the options for the `field-ref` and `assoc` the `base-type`, since it might have `:join-alias` or `:temporal-unit` or other nonsense -- the rules for what `field_ref` should have are really really confusing (I don't even really fully understand them) but all I know is if you break them even slightly it can break legacy viz settings.

We have a legacy MBQL util function for updating legacy ref options somewhere

Reply:
Actually I guess you only need to add `base-type` if the existing field ref uses a string name instead of an integer Field ID